File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/highriskpostcodesupload/RiskyPostCodesMapper.java
package uk.nhs.nhsx.highriskpostcodesupload;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static java.util.Arrays.asList;
import static org.apache.logging.log4j.util.Strings.isBlank;
import static uk.nhs.nhsx.core.exceptions.Validations.throwValidationErrorWith;

public class RiskyPostCodesMapper {

    private static final List<String> VALID_RISK_INDICATORS = asList("H", "M", "L");

    private final Map<String, Map<String, Object>> tierMetadata;
    private final List<String> validTierIndicators;

    public RiskyPostCodesMapper(Map<String, Map<String, Object>> tierMetadata) {
        this.tierMetadata = tierMetadata;
        this.validTierIndicators = new ArrayList<>(tierMetadata.keySet());
    }

    public RiskyPostCodesResult mapOrThrow(RiskyPostDistrictsRequest request) {
        var postCodesToRiskIndicator = new HashMap<String, String>();
        var postCodesToTierIndicator = new HashMap<String, String>();

        request.postDistricts.forEach(
            (postDistrict, indicators) -> {
                validatePostDistrict(postDistrict, indicators);
                postCodesToRiskIndicator.put(postDistrict, indicators.riskIndicator);
                postCodesToTierIndicator.put(postDistrict, indicators.tierIndicator);
            }
        );

        var localAuthoritiesToRiskIndicator = new HashMap<String, String>();
        request.localAuthorities.forEach(
            (localAuthority, indicators) -> {
                validateLocalAuthority(localAuthority, indicators);
                localAuthoritiesToRiskIndicator.put(localAuthority, indicators.tierIndicator);
            }
        );

        var v1 = new RiskyPostCodesV1(postCodesToRiskIndicator);
        var v2 = new RiskyPostCodesV2(postCodesToTierIndicator, localAuthoritiesToRiskIndicator, tierMetadata);

        return new RiskyPostCodesResult(v1, v2);
    }

    private void validatePostDistrict(String postDistrict, PostDistrictIndicators indicators) {
        if (isBlank(postDistrict)) {
            throwValidationErrorWith("Invalid post district - no value");
        }

        if (postDistrict.length() > 20) {
            throwValidationErrorWith("Invalid post district longer than 20 characters: " + postDistrict);
        }

        if (!VALID_RISK_INDICATORS.contains(indicators.riskIndicator)) {
            throwValidationErrorWith(
                "Invalid risk indicator: " + indicators.riskIndicator + " for post district: " + postDistrict
            );
        }

        if (!validTierIndicators.contains(indicators.tierIndicator)) {
            throwValidationErrorWith(
                "Invalid tier indicator: " + indicators.tierIndicator + " for post district: " + postDistrict
            );
        }
    }

    private void validateLocalAuthority(String localAuthority, LocalAuthorityIndicators indicators) {
        if (isBlank(localAuthority)) {
            throwValidationErrorWith("Invalid local authority - no value");
        }

        if (!validTierIndicators.contains(indicators.tierIndicator)) {
            throwValidationErrorWith(
                "Invalid tier indicator: " + indicators.tierIndicator + " for local authority: " + localAuthority
            );
        }
    }

    public String convertToAnalyticsCsv(RiskyPostDistrictsRequest request) {
        return request.postDistricts
            .entrySet().stream()
            .map(toCsvRow())
            .reduce("# postal_district_code, risk_indicator, tier_indicator", String::concat);
    }

    private Function<Map.Entry<String, PostDistrictIndicators>, String> toCsvRow() {
        return postDistrictEntry -> "\n\""
            + postDistrictEntry.getKey() + "\", \""
            + postDistrictEntry.getValue().riskIndicator + "\", \""
            + postDistrictEntry.getValue().tierIndicator + "\"";
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/highriskpostcodesupload/RiskyPostCodesMapper.kt
package uk.nhs.nhsx.highriskpostcodesupload

import uk.nhs.nhsx.core.exceptions.ApiResponseException
import uk.nhs.nhsx.core.exceptions.HttpStatusCode.UNPROCESSABLE_ENTITY_422
import uk.nhs.nhsx.domain.LocalAuthority
import uk.nhs.nhsx.domain.PostDistrict
import uk.nhs.nhsx.domain.PostDistrictIndicators
import uk.nhs.nhsx.domain.RiskIndicator
import uk.nhs.nhsx.domain.TierIndicator

class RiskyPostCodesMapper(private val tierMetadata: Map<String, Map<String, Any>>) {
    private val validTierIndicators = tierMetadata.keys.map { TierIndicator.of(it) }

    fun mapOrThrow(request: RiskyPostDistrictsRequest): RiskyPostCodesResult {
        val postCodesToRiskIndicator = mutableMapOf<PostDistrict, RiskIndicator>()
        val postCodesToTierIndicator = mutableMapOf<PostDistrict, TierIndicator>()
        val localAuthoritiesToRiskIndicator = mutableMapOf<LocalAuthority, TierIndicator>()

        request.postDistricts.forEach { (postDistrict, indicators) ->
            validatePostDistrict(postDistrict, indicators)
            postCodesToRiskIndicator[postDistrict] = indicators.riskIndicator
            postCodesToTierIndicator[postDistrict] = indicators.tierIndicator
        }

        request.localAuthorities.forEach { (localAuthority, indicators) ->
            validateLocalAuthority(localAuthority, indicators)
            localAuthoritiesToRiskIndicator[localAuthority] = indicators.tierIndicator
        }

        val v1 = RiskyPostCodesV1(postCodesToRiskIndicator)
        val v2 = RiskyPostCodesV2(postCodesToTierIndicator, localAuthoritiesToRiskIndicator, tierMetadata)

        return RiskyPostCodesResult(v1, v2)
    }

    private fun validatePostDistrict(postDistrict: PostDistrict, indicators: PostDistrictIndicators) {
        if (!validTierIndicators.contains(indicators.tierIndicator)) {
            throw validationErrorWith(
                "Invalid tier indicator: ${indicators.tierIndicator} for post district: $postDistrict"
            )
        }
    }

    private fun validateLocalAuthority(localAuthority: LocalAuthority, indicators: LocalAuthorityIndicators) {
        if (!validTierIndicators.contains(indicators.tierIndicator)) {
            throw validationErrorWith(
                "Invalid tier indicator: ${indicators.tierIndicator} for local authority: $localAuthority"
            )
        }
    }

    fun convertToAnalyticsCsv(request: RiskyPostDistrictsRequest): String =
        (listOf("# postal_district_code, risk_indicator, tier_indicator") + request.postDistricts
            .map(toCsvRow()))
            .joinToString("\n")

    private fun toCsvRow(): (Map.Entry<PostDistrict, PostDistrictIndicators>) -> String = { it ->
        """"${it.key}", "${it.value.riskIndicator}", "${it.value.tierIndicator}""""
    }

    private fun validationErrorWith(reason: String): ApiResponseException =
        ApiResponseException(UNPROCESSABLE_ENTITY_422, "validation error: $reason")
}
